pepper-apis/src/main/java/org/broadinstitute/ddp/copy/AnswerToAnswerCopier.java

```diff
@@ -0,0 +1,176 @@
+        }
+    }
+
+    public Optional<CompositeQuestionDto> retrieveParentQuestion(QuestionDto childQuestion) {
+        CompositeQuestionDto parentDto = parentDtosByChildId.get(childQuestion.getId());
+        if (parentDto == null) {
+            parentDto = jdbiCompositeQuestion
+                    .findParentDtoByChildQuestionId(childQuestion.getId())
+                    .orElse(null);
+            if (parentDto != null) {
+                for (var childDto : parentDto.getChildQuestions()) {
+                    parentDtosByChildId.put(childDto.getId(), parentDto);
+                }
+            }
+        }
+        return Optional.ofNullable(parentDto);
+    }
+
+    private void copyFromParentAnswer(CompositeAnswer parentAnswer, QuestionDto sourceQuestion,
+                                      FormResponse targetInstance, QuestionDto targetQuestion) {
+        QuestionDto targetParentQuestion = retrieveParentQuestion(targetQuestion)
+                .orElseThrow(() -> new DDPException(
+                        "Could not find parent composite question for target question " + targetQuestion.getStableId()));
+
+        // Find the target composite answer, creating it if it doesn't exist
+        Answer result = targetInstance.getAnswer(targetParentQuestion.getStableId());
+        if (result == null) {
+            CompositeAnswer ans = new CompositeAnswer(null, targetParentQuestion.getStableId(), null);
+            result = answerDao.createAnswer(operatorId, targetInstance.getId(), targetParentQuestion.getId(), ans);
+            targetInstance.putAnswer(result);
+        }
+        CompositeAnswer targetParentAnswer = (CompositeAnswer) result;
+
+        // Line up the rows between source and target composite answers
+        List<AnswerRow> sourceRows = parentAnswer.getValue();
+        List<AnswerRow> targetRows = targetParentAnswer.getValue();
+        while (targetRows.size() < sourceRows.size()) {
+            targetRows.add(new AnswerRow());
+        }
+
+        for (int rowIdx = 0; rowIdx < sourceRows.size(); rowIdx++) {
+            // Line up the columns between source and target rows
+            List<Answer> sourceRow = sourceRows.get(rowIdx).getValues();
+            List<Answer> targetRow = targetRows.get(rowIdx).getValues();
+            while (targetRow.size() < sourceRow.size()) {
+                targetRow.add(null);
+            }
+
+            for (int colIdx = 0; colIdx < sourceRow.size(); colIdx++) {
+                Answer sourceChild = sourceRow.get(colIdx);
+                if (sourceChild == null || !sourceChild.getQuestionStableId().equals(sourceQuestion.getStableId())) {
+                    continue;
+                }
+                Answer targetChild = targetRow.get(colIdx);
+                if (targetChild == null) {
+                    targetChild = createTargetAnswer(sourceChild, targetInstance.getId(), targetQuestion);
+                    jdbiCompositeAnswer.insertChildAnswerItems(
```

Comment:
Kudos to whoever made this `@SqlBatch`'d already--the performance numbers @yufengwng shared look good, so making this sort of optimization way back in the early days of composites was smart :)